### ISSUE TITLE
Keeping validation data out of device memory

### ISSUE THREAD

**Issue by jamesram415** [dmlc/xgboost#11754]
We have quite a large validation set. Training with this full validation set as an eval parameter takes up a large proportion of the device memory that we would ideally use for more training rows. We need metrics on the full val set so can't downsample.

I was curious if there was an existing implementation to essentially do streaming inference over the valset?

I am working on an approach to periodically halt training and loop over the validation set before continuing.

**Comment by trivialfis** [comment]
> I was curious if there was an existing implementation to essentially do streaming inference over the valset?

We don't have this mechanism.

> I am working on an approach to periodically halt training and loop over the validation set before continuing.

Seems like the right direction. The benefit of using the evaluation in XGBoost is that the booster has a prediction cache. Given the same `DMatrix`, it can cache predictions for existing trees, and only run inference for new trees created by the latest training loop. If you are not running evaluations frequently, the slight loss of performance should be acceptable.
